proxy/client.go

```diff
@@ -0,0 +1,169 @@
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/http/httputil"
+
+	"github.com/fsouza/go-dockerclient"
+	. "github.com/weaveworks/weave/common"
+)
+
+type client struct {
+	Interceptor
+	Dial func() (net.Conn, error)
+}
+
+func NewClient(dial func() (net.Conn, error), i Interceptor) *client {
+	return &client{
+		Interceptor: i,
+		Dial:        dial,
+	}
+}
+
+func (c *client) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	req, err := c.InterceptRequest(r)
+	if err == docker.ErrNoSuchImage {
+		http.NotFound(w, r)
+		return
+	} else if err != nil {
+		http.Error(w, "Unable to create proxied request", http.StatusInternalServerError)
+		Warning.Print("Error intercepting request: ", err)
+		return
+	}
+
+	conn, err := c.Dial()
+	if err != nil {
+		http.Error(w, "Could not connect to target", http.StatusInternalServerError)
+		Warning.Print(err)
+		return
+	}
+	client := httputil.NewClientConn(conn, nil)
+	defer client.Close()
+
+	resp, err := client.Do(req)
+	if err != nil && err != httputil.ErrPersistEOF {
+		http.Error(w, fmt.Sprintf("Could not make request to target: %s", err.Error()), http.StatusInternalServerError)
+		Warning.Print("Error forwarding request: ", err)
+		return
+	}
+	resp, err = c.InterceptResponse(resp)
+	if err != nil {
+		http.Error(w, "Unable to intercept response", http.StatusInternalServerError)
+		Warning.Print("Error intercepting response: ", err)
+		return
+	}
+
+	hdr := w.Header()
+	for k, vs := range resp.Header {
+		for _, v := range vs {
+			hdr.Add(k, v)
+		}
+	}
+	Debug.Printf("Response from target: %s %v", resp.Status, w.Header())
+
+	if resp.Header.Get("Content-Type") == RAW_STREAM {
+		doRawStream(w, resp, client)
+	} else if resp.TransferEncoding != nil &&
+		resp.TransferEncoding[0] == "chunked" {
+		doChunkedResponse(w, resp, client)
+	} else {
+		w.WriteHeader(resp.StatusCode)
+		if _, err := io.Copy(w, resp.Body); err != nil {
+			Warning.Print(err)
+		}
+	}
+}
+
+func doRawStream(w http.ResponseWriter, resp *http.Response, client *httputil.ClientConn) {
+	down, downBuf, up, rem, err := hijack(w, client)
+	defer down.Close()
+	defer up.Close()
+
+	if err != nil {
+		Error.Fatal(w, "Unable to hijack connection for raw stream mode", http.StatusInternalServerError)
+		return
+	}
+
+	end := make(chan bool)
+
+	if _, err := down.Write([]byte("HTTP/1.1 200 OK\n")); err != nil {
+		Warning.Print(err)
+		return
+	}
+
+	if err := resp.Header.Write(down); err != nil {
+		Warning.Print(err)
+		return
+	}
+
+	if _, err := down.Write([]byte("\n")); err != nil {
+		Warning.Print(err)
+		return
+	}
+
+	go func() {
+		defer close(end)
+
+		if _, err := io.Copy(down, io.MultiReader(rem, up)); err != nil {
+			Warning.Print(err)
+		}
+	}()
+	go func() {
+		if _, err := io.Copy(up, downBuf); err != nil {
+			Warning.Print(err)
+		}
+		err = up.(interface {
+			CloseWrite() error
+		}).CloseWrite()
+		if err != nil {
+			Debug.Printf("Error Closing upstream: %s", err)
+		}
+	}()
```

Comment:
So this goroutine can stick around beyond the termination of the outer function. Whereas the 'up' goroutine doesn't. is that intentional?